### PR TITLE
[1LP][RFR]Improving test coverage in PXE page

### PR DESCRIPTION
**Pull request by anurag03**
Purpose or Intent
=================
Implementing changes related to copy method and multiple delete in pxe main page.

{{pytest: -v -k 'test_system_image_type_selective_delete or test_customization_template_copy or test_system_image_type_delete_validation or test_customization_template_copy_validation'}}

### CODE REVIEW

File: cfme/infrastructure/pxe.py [hunk 1]
```diff
@@ -574,6 +574,44 @@ def delete(self, cancel=False, *ct_objs):
...
+        if cancel:
+            view.cancel.click()
+            msg = 'Add of new Customization Template was cancelled by the user'
+        else:
+            view.add.click()
+            if self.appliance.version < 5.9:
+                msg = 'Customization Template "{}" was saved'.format(name)
+            else:
+                msg = 'Customization Template "{}" was saved'.format(description)
+        view.flash.assert_success_message(msg)
```
Comment (lkhomenk) [inline cfme/infrastructure/pxe.py]: **REQUIRED**: could you just assert_no_errors() here?
and add separate test that would check flash message after copy

File: cfme/infrastructure/pxe.py [hunk 2]
```diff
@@ -742,6 +780,32 @@ def create(self, name, provision_type, cancel=False):
...
+
+        view.toolbar.configuration.item_select("Remove System Image Types", handle_alert=cancel)
+
+        if not cancel:
+            view.toolbar.configuration.item_select("Remove System Image Types",
+                                                   handle_alert=not cancel)
+            main_view = self.create_view(PXESystemImageTypesView)
+            for sys_obj in sys_objs:
+                msg = 'System Image Type "{}": Delete successful'.format(sys_obj.name)
+                main_view.flash.assert_success_message(msg)
```
Comment (lkhomenk) [inline cfme/infrastructure/pxe.py]: **REQUIRED**: same here - we don't want to get any failures in other tests if we flash messages gets changed.
add separate test for flash messages

File: cfme/tests/infrastructure/test_customization_template.py [hunk 3]
```diff
@@ -21,10 +21,10 @@ def test_customization_template_crud(collection):
                                       image_type='RHEL-6',
                                       script_type='Kickstart',
                                       script_data='Testing the script')
-
+    copy_template_crud = collection.copy(template_crud)
```
Comment (lkhomenk) [inline cfme/tests/infrastructure/test_customization_template.py]: **REQUIRED** do not do like this - don't add everything inside one test_crud - add separate test for customization template copy / delete multiple etc

File: cfme/infrastructure/pxe.py (hunk 1 above)
Comment (anurag03) [reply cfme/infrastructure/pxe.py]: @lkhomenk : I think the correct flash message is part of copy method as the copy operation ends with the flash message.

File: cfme/infrastructure/pxe.py (hunk 2 above)
Comment (anurag03) [reply cfme/infrastructure/pxe.py]: Same as above.

File: cfme/infrastructure/pxe.py (hunk 1 above)
Comment (lkhomenk) [reply cfme/infrastructure/pxe.py]: correct flash message shouldn't block other testing. separate for flash should assert flash message text like this https://github.com/lkhomenk/integration_tests/blob/d54791a439584da3a912b483e6bb417f4c60f84d/cfme/tests/cloud/test_providers.py#L33.

**Comment by dajoRH** [comment]
Lint report for commit f2050e4089bb4a3efb9bbb42244e338910fbd855:

`cfme/tests/infrastructure/test_customization_template.py`:
- :heavy_exclamation_mark: Line 129:1: **W391** *blank line at end of file*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by anurag03** [comment]
@lkhomenk : Done with the suggested changes.

File: cfme/tests/infrastructure/test_customization_template.py [hunk 4]
```diff
@@ -99,3 +99,29 @@ def test_name_max_character_validation(collection):
...
+                                      description=fauxfactory.gen_alphanumeric(16),
+                                      image_type='RHEL-6',
+                                      script_type='Kickstart',
+                                      script_data='Testing the script')
+    copy_template_crud = collection.copy(template_crud)
+    collection.delete(False, copy_template_crud, template_crud)
+
+
+def test_customization_template_copy_validation(appliance, collection):
+    template_crud = collection.create(name=fauxfactory.gen_alphanumeric(8),
```
Comment (lkhomenk) [inline cfme/tests/infrastructure/test_customization_template.py]: REQUIRED: add small doctrings explaining the purpose/intent of the test. that's a musthave right not for new tests

File: cfme/infrastructure/pxe.py [hunk 5]
```diff
@@ -574,6 +574,39 @@ def delete(self, cancel=False, *ct_objs):
...
+         Args:
+             ct_obj : This field is the object of Customization Template.
+             name (str): This field contains the name of the newly copied Customization Template.
+             description (str) : This field contains the description of the newly
+              copied Customization Template.
+             cancel (bool): It's used for flag to cancel or not the copy operation.
+        """
+
+        view = navigate_to(ct_obj, 'Details')
+        view.toolbar.configuration.item_select("Copy this Customization Template")
```
Comment (psav) [inline cfme/infrastructure/pxe.py]: **REQUIRED:** The preference here is to have a new navigation destination I think, that will be linked to the AddView. This way you don't have to go to the trouble of instantiating the view for yourself.

File: cfme/infrastructure/pxe.py [hunk 6]
```diff
@@ -574,6 +574,39 @@ def delete(self, cancel=False, *ct_objs):
...
+        view = ct_obj.create_view(PXECustomizationTemplateAddView)
+        if name is not None:
```
Comment (psav) [inline cfme/infrastructure/pxe.py]: **REQUIRED:** I think the following section should be modified as we skip if the fill is ```None```
```python
        if name is not None:
            view.fill({'name': name})
        if description is not None:
            view.fill({'description': description})
        else:
            description = 'Copy of {}'.format(ct_obj.description)
            view.fill({'description': description})
            name = 'Copy of {}'.format(ct_obj.name)
```
to
```python
        description = description or 'Copy of {}'.format(ct_obj.description)
        view.fill({'name': name, 'description': description})
        name = description
```

File: cfme/infrastructure/pxe.py [hunk 7]
```diff
@@ -574,6 +574,39 @@ def delete(self, cancel=False, *ct_objs):
...
+            view.fill({'name': name})
+        if description is not None:
+            view.fill({'description': description})
+        else:
+            description = 'Copy of {}'.format(ct_obj.description)
+            view.fill({'description': description})
+            name = 'Copy of {}'.format(ct_obj.name)
+
+        customization_templates = self.instantiate(name, description, ct_obj.script_data,
```
Comment (psav) [inline cfme/infrastructure/pxe.py]: This should drop the ```s``` and be singular

File: cfme/infrastructure/pxe.py [hunk 8]
```diff
@@ -742,6 +775,30 @@ def create(self, name, provision_type, cancel=False):
...
+        """
+        This methods deletes the System Image Type using select option,
+         hence can be used for multiple delete.
+
+        Args:
+            cancel: This is the boolean argument required for handle_alert
+            sys_objs: It's System Image Types object
+        """
+        view = navigate_to(self, 'All')
+        rows = view.entities
```
Comment (psav) [inline cfme/infrastructure/pxe.py]: **REQUIRED:** Pretty sure entities has a search or find feature, this shouldn't be necessary

File: cfme/tests/infrastructure/test_system_image_type.py [hunk 9]
```diff
@@ -60,3 +60,36 @@ def test_name_required_error_validation(appliance):
...
+    collection = appliance.collections.system_image_types
+    sys_image_type1 = collection.create(
+        name=fauxfactory.gen_alphanumeric(8),
+        provision_type=SystemImageType.VM_OR_INSTANCE)
+    sys_image_type2 = collection.create(
+        name=fauxfactory.gen_alphanumeric(8),
+        provision_type=SystemImageType.VM_OR_INSTANCE)
+    collection.delete(False, sys_image_type1, sys_image_type2)
+    view = appliance.browser.create_view(PXESystemImageTypesView)
+    msg = 'System Image Type "{}": Delete successful'.format(sys_image_type1.name)
```
Comment (psav) [inline cfme/tests/infrastructure/test_system_image_type.py]: **REQUIRED:** These assertions should be happening inside the delete method, not in the test

Comment (lkhomenk) [reply cfme/tests/infrastructure/test_system_image_type.py]: @psav this test verifies flash text. test above verifies deletion. maybe it would be nice to have helping fixture just to create template

**Comment by dajoRH** [comment]
Lint report for commit 98946bcd5504f6a46f92e7a935d5ba5c557b729d:

`cfme/infrastructure/pxe.py`:
- :bangbang: Line 794:19: **E231** *missing whitespace after ';'*
- :bangbang: Line 794:19: **E702** *multiple statements on one line (semicolon)*

`cfme/tests/infrastructure/test_customization_template.py`:
- :heavy_exclamation_mark: Line 113:64: **W292** *no newline at end of file*

`cfme/tests/infrastructure/test_system_image_type.py`:
- :red_circle: Line 5:1: **F401** *'cfme.infrastructure.pxe.PXESystemImageTypesView' imported but unused*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by anurag03** [comment]
@lkhomenk @psav : done with the suggested changes.

File: cfme/infrastructure/pxe.py [hunk 10]
```diff
@@ -574,6 +580,31 @@ def delete(self, cancel=False, *ct_objs):
             else:
                 navigate_to(ct_obj, 'Details')
 
+    def copy(self, ct_obj, name=None, description=None, cancel=False):
```
Comment (mshriver) [inline cfme/infrastructure/pxe.py]: **Required**

I don't think you should take a `ct_obj` parameter here - the copy operation is against `self`. 

See other framework implementations of entity_object.copy() - they all work against `self`, with some taking `**updates` to override `self` values.

File: cfme/infrastructure/pxe.py [hunk 11]
```diff
@@ -630,7 +670,7 @@ class PXESystemImageTypeDetailsView(PXEMainView):
 
     @View.nested
     class entities(View):  # noqa
-        basic_information = SummaryTable(title="Basic Information")
+        basic_information = Table(title="Basic Information")
```
Comment (mshriver) [inline cfme/infrastructure/pxe.py]: **Required**

I don't see any updates to this widget use to go along with this change and it looks like a SummaryTable to me in the UI, what's the reason for this change?

File: cfme/infrastructure/pxe.py [hunk 12]
```diff
@@ -742,6 +782,26 @@ def create(self, name, provision_type, cancel=False):
...
+         hence can be used for multiple delete.
+
+        Args:
+            cancel: This is the boolean argument required for handle_alert
+            sys_objs: It's System Image Types object
+        """
+        view = navigate_to(self, 'All')
+        for sys_obj in sys_objs:
+            view.entities.row(Name=sys_obj.name)[0].click()
+        view.toolbar.configuration.item_select("Remove System Image Types", handle_alert=cancel)
```
Comment (mshriver) [inline cfme/infrastructure/pxe.py]: **Required**

You've got two calls to configuration.item_select, one before checking cancel and one after.

The call to item_select can be made before checking cancel, as long as `handle_alert` is passed correctly.

I think you should call `flash.assert_no_error` regardless of whether the delete was cancelled.

File: cfme/tests/infrastructure/test_customization_template.py [hunk 13]
```diff
@@ -99,3 +98,16 @@ def test_name_max_character_validation(collection):
...
+def test_customization_template_copy(collection):
+    """
+    Test to check the copy operation of customization templates.
+    """
+    template_crud = collection.create(name=fauxfactory.gen_alphanumeric(8),
+                                      description=fauxfactory.gen_alphanumeric(16),
+                                      image_type='RHEL-6',
+                                      script_type='Kickstart',
+                                      script_data='Testing the script')
+    copy_template_crud = collection.copy(template_crud)
```
Comment (mshriver) [inline cfme/tests/infrastructure/test_customization_template.py]: copy method should be implemented for entity, not collection.

File: cfme/infrastructure/pxe.py (hunk 11 above)
Comment (anurag03) [reply cfme/infrastructure/pxe.py]: I did make use of it, but then changed the code but forgot to change the widget.

File: cfme/infrastructure/pxe.py (hunk 12 above)
Comment (anurag03) [reply cfme/infrastructure/pxe.py]: made this change

**Comment by anurag03** [comment]
@mfalesni can you review this.

**Comment by anurag03** [comment]
@mshriver : Implemented the changes.

**Comment by mshriver** [comment]
Please squash/rebase, only the last commit is verified.

**Comment by anurag03** [comment]
@mshriver  done

File: cfme/infrastructure/pxe.py [hunk 14]
```diff
@@ -500,6 +506,32 @@ def update(self, updates, cancel=False):
...
+         This method is used to copy a Customization Template server via UI.
+
+         Args:
+             name (str): This field contains the name of the newly copied Customization Template.
+             description (str) : This field contains the description of the newly
+              copied Customization Template.
+             cancel (bool): It's used for flag to cancel or not the copy operation.
+        """
+        view = navigate_to(self, 'Copy')
+        name = name or 'Copy of {}'.format(self.name)
```
Comment (mshriver) [inline cfme/infrastructure/pxe.py]: :+1:

File: cfme/infrastructure/pxe.py (hunk 12 above)
Comment (mshriver) [reply cfme/infrastructure/pxe.py]: Close - need to pass `not cancel` to `handle_alert` for item_select.

Check out the docstring for handle_alert:
"handle_alert: How to handle alerts. None - no handling, True - confirm, False - dismiss."
